server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -402,6 +403,39 @@ private void balance() {
         }
     }
 
+    private void warnOnDivergedAllocation() {
+        int total = 0;
+        int undesired = 0;
+
+        for (final var iterator = routingNodes.nodeInterleavedShardIterator(); iterator.hasNext();) {
+            final var shardRouting = iterator.next();
+
+            if (shardRouting.started() == false) {
+                // can only rebalance started shards
+                continue;
+            }
+
+            total++;
+
+            final var assignment = desiredBalance.getAssignment(shardRouting.shardId());
+            if (assignment == null) {
+                // balance is not computed
+                continue;
+            }
+
+            if (assignment.nodeIds().contains(shardRouting.currentNodeId())) {
+                // shard is already on a desired node
+                continue;
+            }
+
+            undesired++;
+        }
+
+        if (total > 100 && undesired > total * 0.1) {
```

Comment:
`0.1` to be converted to a property. Please let me know if you have an idea how to name that fraction.